File: VB/T191208/CustomForm.vb
Imports System.Drawing
Imports System.Windows.Forms
Imports System
Imports DevExpress.LookAndFeel
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Utils

Namespace T191208
    Public Class CustomizationFormEx
        Inherits DevExpress.XtraBars.Customization.CustomizationForm

        'Custom control
        Private control As CustomCustomizationControl
        Protected ReadOnly Property ControlForDrag() As ISupportCustomizationDrag
            Get
                Return TryCast(control, ISupportCustomizationDrag)
            End Get
        End Property
        Protected Overridable ReadOnly Property MyManager() As MyBarManager
            Get
                Return (TryCast(Manager, MyBarManager))
            End Get
        End Property

        Private mouseDownPoint_Renamed As Point
        Protected Overridable Property MouseDownPoint() As Point
            Get
                Return mouseDownPoint_Renamed
            End Get
            Set(ByVal value As Point)
                mouseDownPoint_Renamed = value
            End Set
        End Property

        Private mouseDownItem_Renamed As BarItem
        Protected Overridable Property MouseDownItem() As BarItem
            Get
                Return mouseDownItem_Renamed
            End Get
            Set(ByVal value As BarItem)
                mouseDownItem_Renamed = value
            End Set
        End Property


        Private mouseDraggingItem_Renamed As BarItem
        Protected Overridable Property MouseDraggingItem() As BarItem
            Get
                Return mouseDraggingItem_Renamed
            End Get
            Set(ByVal value As BarItem)
                mouseDraggingItem_Renamed = value
            End Set
        End Property
        Public Sub New(ByVal control As DevExpress.XtraBars.Customization.CustomizationControl, ByVal laf As UserLookAndFeel)
            MyBase.New(control, laf)
            Me.control = TryCast(control, T191208.CustomCustomizationControl)
        End Sub
        Private Sub MouseDownEx(ByVal sender As Object, ByVal e As MouseEventArgs)
            If (e.Button And MouseButtons.Left) <> 0 Then
                MouseDownPoint = System.Windows.Forms.Control.MousePosition
                MouseDownItem = Nothing
                MouseDownItem = ControlForDrag.GetDraggingItem(New Point(e.X, e.Y))
            End If
        End Sub
        Private Sub MouseUpEx(ByVal sender As Object, ByVal e As MouseEventArgs)
            Capture = False
            MouseDownItem = Nothing
        End Sub
        Private Sub MouseMoveEx(ByVal sender As Object, ByVal e As MouseEventArgs)
            If (e.Button And MouseButtons.Left) <> 0 Then
                If MouseDownItem IsNot Nothing Then
                    Dim p As Point = System.Windows.Forms.Control.MousePosition
                    p.Offset(-MouseDownPoint.X, -MouseDownPoint.Y)
                    If Math.Abs(p.X) > 4 OrElse Math.Abs(p.Y) > 4 Then
                        MyManager.SetCustomizeSelectedLink(Nothing)
                        MyManager.SetSelectedItem(MouseDownItem)
                        Dim helper As BarManagerHelpers = MyManager.GetHelper()
                        Try
                            MouseDraggingItem = MouseDownItem
                            Me.ControlForDrag.GetControlToPerformDragging(System.Windows.Forms.Control.MousePosition).AllowDrop = True
                            helper.DragManager.UseDefaultCursors = True
                            helper.DragManager.StartDragging(Nothing, e, MouseDownItem, Me.ControlForDrag.GetControlToPerformDragging(System.Windows.Forms.Control.MousePosition))
                        Finally
                            helper.DragManager.UseDefaultCursors = False
                            Me.ControlForDrag.GetControlToPerformDragging(System.Windows.Forms.Control.MousePosition).AllowDrop = False
                            MyManager.SetSelectedItem(Nothing)
                            MouseDraggingItem = Nothing
                        End Try
                    End If
                End If
            End If
        End Sub

        Public Overrides Sub Init(ByVal AManager As BarManager)
            MyBase.Init(AManager)
            If ControlForDrag IsNot Nothing AndAlso MyManager IsNot Nothing Then
                Me.ControlForDrag.InitControl(AManager)
                ControlForDrag.AddMouseDownHandler(AddressOf MouseDownEx)
                ControlForDrag.AddMouseMoveHandler(AddressOf MouseMoveEx)
                ControlForDrag.AddMouseUpHandler(AddressOf MouseUpEx)
            End If
        End Sub
    End Class

    Public Class MyForm
        Inherits CustomizationFormEx

        Public Sub New(ByVal control As DevExpress.XtraBars.Customization.CustomizationControl, ByVal lookAndFeel As UserLookAndFeel)
            MyBase.New(control, lookAndFeel)
        End Sub
    End Class

End Namespace

File: CS/T191208/CustomForm.cs
using System.Drawing;
using System.Windows.Forms;
using System;
using DevExpress.LookAndFeel;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Utils;

namespace T191208 {
    public class CustomizationFormEx : DevExpress.XtraBars.Customization.CustomizationForm {
        //Custom control
        private CustomCustomizationControl control;
        protected ISupportCustomizationDrag ControlForDrag {
            get {
                return control as ISupportCustomizationDrag;
            }
        }
        protected virtual MyBarManager MyManager {
            get {
                return (Manager as MyBarManager);
            }
        }
        private Point mouseDownPoint;
        protected virtual Point MouseDownPoint {
            get {
                return mouseDownPoint;
            }
            set {
                mouseDownPoint = value;
            }
        }
        private BarItem mouseDownItem;
        protected virtual BarItem MouseDownItem {
            get {
                return mouseDownItem;
            }
            set {
                mouseDownItem = value;
            }
        }

        private BarItem mouseDraggingItem;
        protected virtual BarItem MouseDraggingItem {
            get {
                return mouseDraggingItem;
            }
            set {
                mouseDraggingItem = value;
            }
        }
        public CustomizationFormEx(DevExpress.XtraBars.Customization.CustomizationControl control, UserLookAndFeel laf)
            : base(control, laf) {
            this.control = control as T191208.CustomCustomizationControl;
        }
        void MouseDownEx(object sender, MouseEventArgs e) {
            if ((e.Button & MouseButtons.Left) != 0) {
                MouseDownPoint = Control.MousePosition;
                MouseDownItem = null;
                MouseDownItem = ControlForDrag.GetDraggingItem(new Point(e.X, e.Y));
            }
        }
        void MouseUpEx(object sender, MouseEventArgs e) {
            Capture = false;
            MouseDownItem = null;
        }
        void MouseMoveEx(object sender, MouseEventArgs e) {
            if ((e.Button & MouseButtons.Left) != 0) {
                if (MouseDownItem != null) {
                    Point p = Control.MousePosition;
                    p.Offset(-MouseDownPoint.X, -MouseDownPoint.Y);
                    if (Math.Abs(p.X) > 4 || Math.Abs(p.Y) > 4) {
                        MyManager.SetCustomizeSelectedLink(null);
                        MyManager.SetSelectedItem(MouseDownItem);
                        BarManagerHelpers helper = MyManager.GetHelper();
                        try {
                            MouseDraggingItem = MouseDownItem;
                            this.ControlForDrag.GetControlToPerformDragging(Control.MousePosition).AllowDrop = true;
                            helper.DragManager.UseDefaultCursors = true;
                            helper.DragManager.StartDragging(null, e, MouseDownItem, this.ControlForDrag.GetControlToPerformDragging(Control.MousePosition));
                        }
                        finally {
                            helper.DragManager.UseDefaultCursors = false;
                            this.ControlForDrag.GetControlToPerformDragging(Control.MousePosition).AllowDrop = false;
                            MyManager.SetSelectedItem(null);
                            MouseDraggingItem = null;
                        }
                    }
                }
            }
        }

        public override void Init(BarManager AManager) {
            base.Init(AManager);
            if (ControlForDrag != null && MyManager != null) {
                this.ControlForDrag.InitControl(AManager);
                ControlForDrag.AddMouseDownHandler(MouseDownEx);
                ControlForDrag.AddMouseMoveHandler(MouseMoveEx);
                ControlForDrag.AddMouseUpHandler(MouseUpEx);
            }
        }
    }

    public class MyForm : CustomizationFormEx {
        public MyForm(DevExpress.XtraBars.Customization.CustomizationControl control, UserLookAndFeel lookAndFeel)
            : base(control, lookAndFeel) {
        }
    }
    
}
